File: sdk/src/main/java/com/cxense/cxensesdk/RawJsonAdapter.java
package com.cxense.cxensesdk;

import com.google.gson.TypeAdapter;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class RawJsonAdapter extends TypeAdapter<List<String>> {
    @Override
    public void write(JsonWriter out, List<String> value) throws IOException {
        out.beginArray();
        for (String s : value) {
            out.jsonValue(s);
        }
        out.endArray();
    }

    @Override
    public List<String> read(JsonReader in) throws IOException {
        // not optimized version
        List<String> data = new ArrayList<>();
        in.beginArray();
        while (in.hasNext())
            data.add(in.nextString());
        in.endArray();
        return data;
    }
}

File: sdk/src/main/java/com/cxense/cxensesdk/RawJsonAdapter.kt
package com.cxense.cxensesdk

import androidx.annotation.RestrictTo
import com.google.gson.TypeAdapter
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonWriter

/**
 * Used for writing prepared json as raw string.
 *
 */
@RestrictTo(RestrictTo.Scope.LIBRARY)
internal class RawJsonAdapter : TypeAdapter<List<String>>() {
    override fun write(output: JsonWriter, value: List<String>?) {
        with(output) {
            beginArray()
            value?.forEach {
                jsonValue(it)
            }
            endArray()
        }
    }

    override fun read(input: JsonReader): List<String> {
        // not optimized version
        return input.readList { it.nextString() }
    }
}
